libcudacxx/include/cuda/__ptx/instructions/shfl.h

```diff
@@ -0,0 +1,136 @@
+//
+//===----------------------------------------------------------------------===//
+
+#if !defined(_CUDA_PTX_SHFL_SYNC_H)
+#  define _CUDA_PTX_SHFL_SYNC_H
+
+#  include <cuda/std/detail/__config>
+
+#  if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#    pragma GCC system_header
+#  elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#    pragma clang system_header
+#  elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#    pragma system_header
+#  endif // no system header
+
+#  if _CCCL_STD_VER >= 2017
+
+#    include <cuda/__ptx/instructions/get_sreg.h>
+#    include <cuda/__ptx/ptx_dot_variants.h>
+#    include <cuda/std/__type_traits/is_integral.h>
+#    include <cuda/std/__type_traits/is_signed.h>
+#    include <cuda/std/cstdint>
+
+#    include <nv/target> // __CUDA_MINIMUM_ARCH__ and friends
+
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA_PTX
+
+#    if __cccl_ptx_isa >= 600
+
+struct shfl_return_values
+{
+  _CUDA_VSTD::uint32_t __data;
+  _CUDA_VSTD::int32_t __pred;
+};
+
+template <dot_shfl_mode _ShuffleMode>
+_CCCL_NODISCARD _CCCL_DEVICE static inline shfl_return_values shfl_sync(
+  shfl_mode_t<_ShuffleMode> __shfl_mode,
+  _CUDA_VSTD::uint32_t __data,
+  _CUDA_VSTD::uint32_t __lane_idx_offset,
+  _CUDA_VSTD::uint32_t __clamp_segmask,
+  _CUDA_VSTD::uint32_t __lane_mask) noexcept
+{
+  _CCCL_ASSERT(__lane_idx_offset < 32, "the lane index or offset must be less than the warp size");
+  _CCCL_ASSERT(__clamp_segmask <= 0b111111111111, "clamp value + segmentation mask must be less or equal than 12 bits");
+  _CCCL_ASSERT((__lane_mask & __activemask()) == __lane_mask, "lane mask must be a subset of the active mask");
+#      if defined(_CCCL_ENABLE_DEBUG_MODE)
+  auto __lane                = get_sreg_laneid();
+  auto __clamp               = __clamp_segmask & 0b11111;
+  auto __segmask             = __clamp_segmask >> 8;
+  auto __max_lane            = (__lane & __segmask) | (__clamp & __segmask);
+  _CUDA_VSTD::uint32_t __dst = 0;
+#      endif
+  _CUDA_VSTD::int32_t __pred;
+  _CUDA_VSTD::uint32_t __ret;
+  if constexpr (__shfl_mode == shfl_mode_idx)
+  {
+#      if defined(_CCCL_ENABLE_DEBUG_MODE)
+    auto __min_lane = (__lane & __segmask);
+    auto __j        = __min_lane | (__lane_idx_offset & ~__segmask);
+    __dst           = __j <= __max_lane ? __j : __lane;
+#      endif
+    asm volatile(
+      "{                                                           \n\t\t"
+      ".reg .pred p;                                               \n\t\t"
+      "shfl_sync.sync.idx.b32 %0|p, %2, %3, %4, %5;                \n\t\t"
+      "selp.s32 %1, 1, 0, p;                                         \n\t"
+      "}"
+      : "=r"(__ret), "=r"(__pred)
+      : "r"(__data), "r"(__lane_idx_offset), "r"(__clamp_segmask), "r"(__lane_mask));
+  }
+  else if constexpr (__shfl_mode == shfl_mode_up)
+  {
+#      if defined(_CCCL_ENABLE_DEBUG_MODE)
+    auto __j = __lane - __lane_idx_offset;
+    __dst    = __j >= __max_lane ? __j : __lane;
+#      endif
```

Comment:
I am not a fan of those conditionals, could we rather define a function that does that and call that inside of the _CCCL_ASSERT?